<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Qstats</title>
<link rev="made" href="mailto:root@hampsten.nonet" />
</head>

<body style="background-color: white">

<p><a name="__index__"></a></p>
<!-- INDEX BEGIN -->
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#input_format">INPUT FORMAT</a></li>
	<li><a href="#output">OUTPUT</a></li>
	<li><a href="#examples">EXAMPLES</a></li>
	<li><a href="#references">REFERENCES</a></li>
	<li><a href="#bugs">BUGS</a></li>
	<li><a href="#see_also">SEE ALSO</a></li>
	<li><a href="#contact_info">CONTACT INFO</a></li>
</ul>
-->
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>Qstats - Calculate basic statistics  for a QTL dataset.</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p><strong>Qstats</strong> [ <strong>-o</strong> <em>output</em> ] [ <strong>-i</strong> <em>input</em> ] [ <strong>-m</strong> <em>mapfile</em> ]  [ <strong>-p</strong> <em>yes</em> ]</p>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p><strong>Qstats</strong> does some basic statistics on a dataset of quantitative traits. 
It plots a histogram and calculates the sample size, mean, variance
standard deviation, skewness, kurtosis, and average deviation for a 
quantitative trait.   The program also summarizes missing marker and trait data, as will as 
determining the marker types (dominant or codominant).  Finally, <strong>Qstats</strong> will
test whether markers are segregating at random. 
It requires a molecular map that could be a random one produced by 
<strong>Rmap</strong>,  or a real one in the same format as the output of 
<strong>Rmap</strong>.   The sample could be a randomly generated one from 
<strong>Rcross</strong> or a real one in the same format as the output of 
<strong>Rcross</strong>.</p>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>See <strong>QTLcart(1)</strong> for more information on the global options
<strong>-h</strong> for help, <strong>-A</strong> for automatic,  <strong>-V</strong> for non-Verbose
<strong>-W path</strong> for a working directory, <strong>-R file</strong> to specify a resource
file, <strong>-e</strong> to specify the log file, <strong>-s</strong> to specify a seed for the
random number generator and <strong>-X stem</strong> to specify a filename stem. 
The options below are specific to this program.</p>
<p>If you use this program without specifying any options, then you will
get into a menu that allows you to set them interactively.</p>
<dl>
<dt><strong><a name="item__2do"><strong>-o</strong></a></strong><br />
</dt>
<dd>
This requires a filename for output.   <strong>Qstats</strong> will append the file if
it exists, and create a new file if it does not.   If not used, then <strong>Qstats</strong> will use
<em>qtlcart.qst</em>.
</dd>
<p></p>
<dt><strong><a name="item__2di"><strong>-i</strong></a></strong><br />
</dt>
<dd>
This requires an input filename.    This file must exist.  It should be in the
same format as the output of <strong>Rcross</strong>.  The default file is <em>qtlcart.cro</em>.
</dd>
<p></p>
<dt><strong><a name="item__2dm"><strong>-m</strong></a></strong><br />
</dt>
<dd>
<strong>Qstats</strong> requires a genetic linkage map.  This option require
the name of a file containing the map.  It should be in the same format
that <strong>Rmap</strong> outputs.  The default file is <em>qtlcart.map</em>.
</dd>
<p></p>
<dt><strong><a name="item__2dp"><strong>-p</strong></a></strong><br />
</dt>
<dd>
requires an argument <em>yes</em> or <em>no</em>.   By default it is <em>no</em>.   If used
with <em>yes</em>, then the probability distribution for each marker and each individual
will be written in a section at the end of the output file.
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="input_format">INPUT FORMAT</a></h1>
<p>The input format of the molecular map should be the same as that of the output 
format from the program 
<strong>Rmap</strong>.   The input format of the individual data should be the same as the output format
of the program 
<strong>Rcross</strong>.</p>
<p>
</p>
<hr />
<h1><a name="output">OUTPUT</a></h1>
<p>As mentioned above, the output will have sections for basic statistics on the traits,
missing data summaries and segregation statistics.   If the <strong>-p</strong> option is used
with an argument <em>yes</em>, then marker distributions are printed in a section at the
end.   For each marker, there will be a table that has individuals as rows and
the probability distribution and expected values of the marker as columns.
An example of the output is</p>
<pre>
    -begin markerprobs
    -markername BD267                 -chromosome     1 -marker     1
    -h   ind.   p(QQ)   p(Qq)    p(qq)   E(a)     E(d)
            1  0.3304  0.5242  0.1454   0.1851   0.0242
            2  0.3441  0.4850  0.1709   0.1731  -0.0150
            3  0.3747  0.4958  0.1295   0.2452  -0.0042
            4  0.0696  0.3884  0.5421  -0.4725  -0.1116
            5  0.2597  0.5696  0.1707   0.0889   0.0696
            6  0.2903  0.5500  0.1598   0.1305   0.0500
            7  0.2903  0.5500  0.1598   0.1305   0.0500
            8  0       0       1       -1.0     -0.5</pre>
<p>The section begins with <em>-begin markerprobs</em> and ends with <em>-end markerprobs</em>.  Each marker
is then listed and a header line is followed by the output.   The first column is the individual
number.  The next three are the probabilities of QQ, Qq and qq genotypes, respectively.
Then, the expected values for additive and dominance effects are calculated via</p>
<pre>
    E(a) = p(QQ) - p(qq)
    E(d) = [p(Qq)-p(QQ)-p(qq)]/2</pre>
<p>One could use the final two values as replacements for the marker values in a regression 
analysis.   The additive expectation is calculated assuming that QQ genotypes have value 1,
while Qq and qq genotypes have values 0 and -1, respectively.  The dominance expectation
sets heterozygotes equal to 1/2 and homozygotes to -1/2.</p>
<p>This option produces a lot of output: It is off by default.   The probabilities follow from
Jiang and Zeng (1997).</p>
<p>Note that the line for individual 8 has <code>p(QQ)</code> = <code>p(Qq)</code> = 0, while <code>p(qq)</code> = 1.
This is a dominant marker and individual 8 had the unambiguous genotype qq.  
You would see similar output for codominant markers.</p>
<p>
</p>
<hr />
<h1><a name="examples">EXAMPLES</a></h1>
<pre>
    % Qstats -i corn.cro -m corn.map</pre>
<p>Calculates basic statistics on the dataset in <em>corn.cro</em> using the genetic linkage map 
in <em>corn.map</em>.  The program will display and interactive menu for setting
options and print out messages to the screen while running.  These
can be turned off with <strong>-A</strong> and <strong>-V</strong>, respectively.  If the dataset in 
<em>corn.cro</em> has more than one trait, then all traits will be analyzed.</p>
<p>
</p>
<hr />
<h1><a name="references">REFERENCES</a></h1>
<ol>
<li></li>
M. Lynch and B. Walsh (1998) Genetics and Analysis of Quantitative Traits.  Sinauer Associates,
Sunderland, MA.
<p></p>
<li></li>
C. Jiang and Z.-B. Zeng (1997) Mapping quantitative trait loci with dominant and missing markers in 
various crosses from two inbred lines.  <em>Genetica</em> <strong>101</strong>, 47-58.
<p></p></ol>
<p>
</p>
<hr />
<h1><a name="bugs">BUGS</a></h1>
<p>Are there any other statistics that we can do?  Your suggestions are welcome.</p>
<p>
</p>
<hr />
<h1><a name="see_also">SEE ALSO</a></h1>
<p><strong>Emap(1)</strong>,  
<strong>Rmap(1)</strong>,  
<strong>Rqtl(1)</strong>, 
<strong>Rcross(1)</strong>, 
<strong>Qstats(1)</strong>, 
<strong>LRmapqtl(1)</strong>,
<strong>BTmapqtl(1)</strong>,
<strong>SRmapqtl(1)</strong>, 
<strong>JZmapqtl(1)</strong>, 
<strong>Eqtl(1)</strong>,
<strong>Prune(1)</strong>, 
<strong>Preplot(1)</strong>,  
<strong>MImapqtl(1)</strong>, 
<strong>MultiRegress(1)</strong>,
<strong>Examples(1)</strong>
<strong>SSupdate.pl(1)</strong>, 
<strong>Prepraw.pl(1)</strong>, 
<strong>EWThreshold.pl(1)</strong>, 
<strong>GetMaxLR.pl(1)</strong>, 
<strong>Permute.pl(1)</strong>, 
<strong>Vert.pl(1)</strong>, 
<strong>CWTupdate.pl(1)</strong>, 
<strong>Ztrim.pl(1)</strong>, 
<strong>SRcompare.pl(1)</strong>, 
<strong>Ttransform.pl(1)</strong>, 
<strong>TestExamples.pl(1)</strong>, 
<strong>Model8.pl(1)</strong>, 
<strong>Dobasics.pl(1)</strong>, 
<strong>Bootstrap.pl(1)</strong></p>
<p>
</p>
<hr />
<h1><a name="contact_info">CONTACT INFO</a></h1>
<p>In general, it is best to contact us via email (<a href="mailto:basten@statgen.ncsu.edu">basten@statgen.ncsu.edu</a>)</p>
<pre>
        Christopher J. Basten, B. S. Weir and Z.-B. Zeng
        Bioinformatics Research Center, North Carolina State University
        1523 Partners II Building/840 Main Campus Drive
        Raleigh, NC 27695-7566     USA
        Phone: (919)515-1934</pre>
<p>Please report all bugs via email to <a href="mailto:qtlcart-bug@statgen.ncsu.edu.">qtlcart-bug@statgen.ncsu.edu.</a></p>
<p>The <strong>QTL Cartographer</strong> web site ( <a href="http://statgen.ncsu.edu/qtlcart">http://statgen.ncsu.edu/qtlcart</a> ) has
links to the manual, man pages, ftp server and supplemental 
materials.</p>

</body>

</html>
