<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Zmapqtl</title>
<link rev="made" href="mailto:root@hampsten.nonet" />
</head>

<body style="background-color: white">

<p><a name="__index__"></a></p>
<!-- INDEX BEGIN -->
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#input_format">INPUT FORMAT</a></li>
	<li><a href="#examples">EXAMPLES</a></li>
	<li><a href="#models">MODELS</a></li>
	<li><a href="#permutation_tests">PERMUTATION TESTS</a></li>
	<li><a href="#references">REFERENCES</a></li>
	<li><a href="#caveats">CAVEATS</a></li>
	<li><a href="#bugs">BUGS</a></li>
	<li><a href="#see_also">SEE ALSO</a></li>
	<li><a href="#contact_info">CONTACT INFO</a></li>
</ul>
-->
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>Zmapqtl - Composite interval mapping module</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p><strong>Zmapqtl</strong> [ <strong>-o</strong> <em>output</em> ] [ <strong>-i</strong> <em>input</em> ] [ <strong>-m</strong> <em>mapfile</em> ]
[ <strong>-l</strong> <em>lrfile</em> ] [ <strong>-S</strong> <em>srfile</em> ] [ <strong>-t</strong> <em>trait</em> ] [ <strong>-M</strong> <em>Model</em> ]  
[ <strong>-c</strong> <em>chrom</em> ] [ <strong>-d</strong> <em>walk</em> ] [ <strong>-n</strong> <em>nbp</em> ] [ <strong>-w</strong> <em>window</em> ]
[ <strong>-r</strong> <em>perms</em> ]  [ <strong>-r</strong> <em>boots</em> ]</p>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p><strong>Zmapqtl</strong> uses composite interval mapping to map quantitative trait loci to
a map of molecular markers. It requires a molecular map that could be a random one produced by 
<strong>Rmap</strong>,  or a real one in the same format as the output of 
<strong>Rmap</strong>.   The sample could be a randomly generated one from 
<strong>Rcross</strong> or a real one in the same format as the output of 
<strong>Rcross</strong>. In addition, the program requires the results of the stepwise linear regression 
analysis of  <strong>SRmapqtl</strong> for composite interval mapping.</p>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>See <strong>QTLcart(1)</strong> for more information on the global options
<strong>-h</strong> for help, <strong>-A</strong> for automatic,  <strong>-V</strong> for non-Verbose
<strong>-W path</strong> for a working directory, <strong>-R file</strong> to specify a resource
file, <strong>-e</strong> to specify the log file, <strong>-s</strong> to specify a seed for the
random number generator and <strong>-X stem</strong> to specify a filename stem. 
The options below are specific to this program.</p>
<p>If you use this program without specifying any options, then you will
get into a menu that allows you to set them interactively.</p>
<dl>
<dt><strong><a name="item__2do"><strong>-o</strong></a></strong><br />
</dt>
<dd>
This requires a filename for output.   <strong>Zmapqtl</strong> will append the file if
it exists, and create a new file if it does not.   If not used, then <strong>Zmapqtl</strong> will use
<em>qtlcart.z</em>.
</dd>
<p></p>
<dt><strong><a name="item__2di"><strong>-i</strong></a></strong><br />
</dt>
<dd>
This requires an input filename.    This file must exist.  It should be in the
same format as the output of <strong>Rcross</strong>. The default file is <em>qtlcart.cro</em>.
</dd>
<p></p>
<dt><strong><a name="item__2dm"><strong>-m</strong></a></strong><br />
</dt>
<dd>
<strong>Zmapqtl</strong> requires a genetic linkage map.  This option requires
the name of a file containing the map.  It should be in the same format
that <strong>Rmap</strong> outputs.  The default file is <em>qtlcart.map</em>.
</dd>
<p></p>
<dt><strong><a name="item__2dt"><strong>-t</strong></a></strong><br />
</dt>
<dd>
Use this to specify which trait <strong>Zmapqtl</strong>
will analyze.  If this number is greater than
the number of traits, then all traits will be analyzed.   The default is
to analyze trait 1 only.
</dd>
<p></p>
<dt><strong><a name="item__2dl"><strong>-l</strong></a></strong><br />
</dt>
<dd>
Allows the user to specify the name of the file containing results from
<strong>LRmapqtl</strong>.   <strong>Zmapqtl</strong> reads those results and uses the information to
choose cofactors for some of the analysis methods.
</dd>
<p></p>
<dt><strong><a name="item__2ds"><strong>-S</strong></a></strong><br />
</dt>
<dd>
Allows the user to specify the name of the file containing results from
<strong>SRmapqtl</strong>. <strong>Zmapqtl</strong> reads the results and uses the information to choose
cofactors for composite interval mapping model 6.
</dd>
<p></p>
<dt><strong><a name="item__2dm"><strong>-M</strong></a></strong><br />
</dt>
<dd>
<strong>Zmapqtl</strong> assumes the specified model (see below) in the analysis. Model 3 is default.
</dd>
<p></p>
<dt><strong><a name="item__2dc"><strong>-c</strong></a></strong><br />
</dt>
<dd>
The user can specify a specific chromosome for <strong>Zmapqtl</strong> to analyze.   
If zero, then all will be analyzed.
</dd>
<p></p>
<dt><strong><a name="item__2dd"><strong>-d</strong></a></strong><br />
</dt>
<dd>
<strong>Zmapqtl</strong> walks along the chromosome at a rate that can be specified with this
option.  The default is to do an analysis
every  2 centiMorgans along the chromosome.
</dd>
<p></p>
<dt><strong><a name="item__2dn"><strong>-n</strong></a></strong><br />
</dt>
<dd>
Use this to indicate how many background parameters <strong>Zmapqtl</strong>
uses  in composite interval  mapping.
This is used only with model 6, and gives an upper bound.  If fewer than
this number of markers are ranked in the <em>SRmapqtl.out</em> file, then
less than the specified number of markers will be used.
</dd>
<p></p>
<dt><strong><a name="item__2dw"><strong>-w</strong></a></strong><br />
</dt>
<dd>
<strong>Zmapqtl</strong>
blocks out a region of this many centiMorgans on either side of the markers flanking the test 
position when picking background markers.  It is 10 by default and is only used in
models 5 and 6.  We refer to it as the <em>window size</em>.
</dd>
<p></p>
<dt><strong><a name="item__2dr"><strong>-r</strong></a></strong><br />
</dt>
<dd>
<strong>Zmapqtl</strong> can do a permutation test to determine the threshold for rejecting the
null hypothesis of no QTL at a site.  By default, this option sets the
number of permutations equal to 0, which means no permutation test is
run.  You can set it to a number &lt; 10000 to do the test.  See Churchill and
Doerge (1994) for more details.  The results are in an interim file.  Use
<strong>Eqtl</strong> to summarize them when enough repetitions have been done.  You need to run
<strong>Zmapqtl</strong> without permutations or bootstraps at least once before you can do the 
permutation tests.  This option only allows for interval mapping (Model 3) or
composite interval mapping (Model 6).
</dd>
<p></p>
<dt><strong><a name="item__2db"><strong>-b</strong></a></strong><br />
</dt>
<dd>
When used with argument 1, <strong>Zmapqtl</strong> will do a single bootstrap.  You need to run 
<strong>Prune</strong>  to actually create the bootstrapped data set:  This option merely 
analyzes it and stores summary statistics in an interim file
<em>qtlcart.z3b</em> by default, for model 3.  You should also run 
<strong>Zmapqtl</strong>  without bootstraps or
permutation tests before doing a bootstrap analysis.
</dd>
<dd>
<p>When used with an argument 2, <strong>Zmapqtl</strong> will do a jackknife analysis.  Again,
<strong>Zmapqtl</strong> should be run without this argument prior to doing a jackknife.</p>
</dd>
<p></p></dl>
<p>
</p>
<hr />
<h1><a name="input_format">INPUT FORMAT</a></h1>
<p>The input format of the molecular map should be the same as that of the output 
format from the program 
<strong>Rmap</strong>.   The input format of the individual data should be the same as the output format
of the program 
<strong>Rcross</strong>.</p>
<p>
</p>
<hr />
<h1><a name="examples">EXAMPLES</a></h1>
<pre>
        % Zmapqtl</pre>
<p>Calculates the likelihood ratio test statistics of the dataset in <em>qtlcart.cro</em>
using the map in <em>qtlcart.map</em>.</p>
<pre>
        % nice Zmapqtl -A -V -i corn.cro -m corn.map -M 6 -r 500 &amp;</pre>
<p>Calculates the likelihood ratio test statistics of the dataset in <em>corn.cro</em> 
using the map in <em>corn.map</em>.  Model 6 is used for analysis and a permutation test
with 500 replications is performed.  The program is nice'd as a courtesy to other
users, and run in the background so that the user can logout and relax.</p>
<p>
</p>
<hr />
<h1><a name="models">MODELS</a></h1>
<p>Different parameters for the <strong>-M</strong> option allow for the analysis of the data
assuming different models.  Models 1-3 were described in Zeng (1993, 1994).</p>
<ol>
<li></li>
Fit all the background markers.  This was meant for illustration of the method in
the original paper (Zeng, 1993), and should not be used for analysis.
<p></p>
<li></li>
Fit all unlinked background markers.  This is another illustrative example and also 
should not be used.
<p></p>
<li></li>
Fit only the mean (Lander and Botstein (1989) method)
<p></p>
<li></li>
Fit a subset of the other markers, namely those unlinked markers with the 
highest correlation with the trait on each chromosome.  This is an <em>ad hoc</em> model
programmed in anticipation of model 6 below.
<p></p>
<li></li>
This model uses a pair of markers from each other chromosome and all linked 
markers that fall
outside a window around the flanking markers.  This window extends to 10 cM 
beyond the 
markers immediately flanking the test position.  The window size can be changed with
the <strong>-w</strong> option.  This is another <em>ad hoc</em> model
programmed in anticipation of model 6 below.
<p></p>
<li></li>
This model uses a specified number of markers that fall
outside a window around the flanking markers.  This window extends to 10 cM 
beyond the 
markers immediately flanking the test position.  The number of markers are 
set by the <strong>-n</strong> option.  You need to run  <strong>SRmapqtl</strong>
to rank the markers before using model 6.    You should use this for composite 
interval mapping when using markers ranked by <strong>SRmapqtl</strong>.
<p></p>
<li></li>
This model requires that you have already run <strong>Zmapqtl</strong> and <strong>Eqtl</strong>.  It reads in
the estimated QTL from the <em>Eqtl.out</em> file and uses them as virtual markers to control for
the genetic background.   All identified markers are used that do not fall within the
window.   This has not been extensively tested:  Use it at your own risk.
<p></p>
<li></li>
Like model 7, this requires a prior run of <strong>Zmapqtl</strong> and <strong>Eqtl</strong>.  Instead of using 
virtual markers, <strong>Zmapqtl</strong> uses the closest flanking markers to identified QTL.  
Again, all of these markers outside the window are used.   This is good model.
You can iterate this process to see if a stable set of cofactors can be identified.
See the script below and the <strong>Model8.pl(1)</strong> man page for more details.
<p></p></ol>
<p>The default is to fit only the mean, that is to use interval mapping.</p>
<p>
</p>
<hr />
<h1><a name="permutation_tests">PERMUTATION TESTS</a></h1>
<p>Churchill and Doerge (1994) describe a method to calculate the threshold
values for quantitative trait mapping that we have implemented in this
program.  Basically, it does a permutation of the trait values and the
genotypes and redoes the analysis.  Over the number of replicates, two
types of thresholds are defined: ``experimentwise'' and ``comparisonwise''.
We calculate the experimentwise thresholds, but only give p values
for the comparisonwise values to save on storage space.  The p values
give the proportion of permuted replicates that have loglikelihood 
ratios larger than the observed ratios.</p>
<p>If you choose to do permutation tests, you need to run  <strong>Zmapqtl</strong>
with the model of choice prior to doing the permutation test.  Also, if 
the program terminates prematurely, you can restart it from where it left
off to complete the permutation test.</p>
<p>
</p>
<hr />
<h1><a name="references">REFERENCES</a></h1>
<ol>
<li></li>
Churchill, G. A.  and R. W. Doerge (1994) Empirical threshold values for 
quantitative trait mapping. <em>Genetics</em> <strong>138</strong>, 963-971.
<p></p>
<li></li>
Lander, E. S. and D. Botstein (1989) Mapping Mendelian factors underlying 
quantitative traits using RFLP linkage maps. <em>Genetics</em> <strong>121</strong>, 185-199.
<p></p>
<li></li>
Zeng, Zhao-Bang (1993) Theoretical basis for separation of multiple linked gene 
effects in mapping quantitative trait loci.
<em>Proc. Natl. Acad. Sci., USA</em> <strong>90</strong>, 10972-10976.
<p></p>
<li></li>
Zeng, Zhao-Bang (1994) Precision mapping of quantitative trait loci.
<em>Genetics</em> <strong>136</strong>, 1457-1468.
<p></p></ol>
<p>
</p>
<hr />
<h1><a name="caveats">CAVEATS</a></h1>
<p>Model 7 has not been extensively tested.   Remember that you will need to run
<strong>Zmapqtl</strong> with some other model (say 3 or 6) and then run <strong>Eqtl</strong> prior to using
model 7.</p>
<p>Model 8 requires a prior run of <strong>Zmapqtl</strong> and <strong>Eqtl</strong> as well.   There is a shell script
called <strong>Model8.csh</strong> in the <strong>scripts</strong> subdirectory that allows you to easily iterate using
model 8.   Try using it to see whether your set of cofactors (or estimated QTL) becomes
stable after a few iterations.   A low threshold may lead to adding and dropping putative
QTL over consecutive analyses using Model 8.  Here is the shell script:</p>
<pre>
        #!/usr/bin/csh
        #
        #  Run Model 8 iteration
        #           Copyright 2001 Christopher J. Basten
        #   Usage:
        #   Model8  bindir  stem   siglevel  iterations  max_nbp
        #     bindir is the binary subdirectory
        #     stem is the filename stem
        #     siglevel  is the significance level to declare a QTL
        #     iterations is the number of iterations
        #     max_nbp is the maximal number of background parameters.
        #
        if ( $1 == '-h' ) then
        echo &quot;    Usage:  Model8.csh bindir  stem   siglevel  iterations  max_nbp&quot;
        echo &quot;Where&quot;
        echo &quot;        bindir  = QTL Cart. binary directory&quot;
        echo &quot;          stem  = filename stem&quot;
        echo &quot;      siglevel  = Significance level to declare a QTL&quot;
        echo &quot;    iterations  = number of iterations&quot;
        echo &quot;       max_nbp  = maximal number of background parameters&quot;
        echo &quot; &quot;
        echo &quot;Now exiting&quot;
        exit
        endif
        set bindir=$1
        set stem=$2
        set siglevel=$3
        set iterations=$4
        set maxnbp=$5
        $bindir/Qstats -X $stem -A -V
        $bindir/Zmapqtl -A -V -M 3
        $bindir/Eqtl -A -V -S $siglevel
        #
        #  Save the original files
        #
        /usr/bin/mv $stem.eqt $stem.eqt.0
        /usr/bin/mv $stem.z $stem.z.0
        /usr/bin/cp $stem.sr $stem.sr.0
        #
        #  Use model 8 iteratively with cofactors from previous run.
        #
        set i=1
        while ( $i &lt; $iterations )
        echo &quot;Doing iteration $i&quot;
        $bindir/Zmapqtl -A -V -M 8 -n $maxnbp
        /usr/bin/rm $stem.sr 
        $bindir/Eqtl -A -V -S $siglevel
        /usr/bin/cp $stem.sr $stem.sr.$i
        /usr/bin/mv $stem.eqt $stem.eqt.$i
        /usr/bin/mv $stem.z $stem.z.$i
        @ i++
        end
        /usr/bin/rm $stem.sr 
        echo &quot;Finished&quot;</pre>
<p>The above script has been translated into <strong>Perl</strong> to make it more useful.  The 
<strong>Model8.pl</strong> script can take command line parameters and is self-documenting.</p>
<p>
</p>
<hr />
<h1><a name="bugs">BUGS</a></h1>
<p>It is likely that we will abandon the internal permutation tests in <strong>Zmapqtl</strong>.  It is
more efficient to use <strong>Prune</strong> and a batch file to do the same job.  This paradigm will 
allow users to do permutation tests with any of the programs.  Of course, you will
need access to a UNIX platform to do this.</p>
<p>
</p>
<hr />
<h1><a name="see_also">SEE ALSO</a></h1>
<p><strong>Emap(1)</strong>,  
<strong>Rmap(1)</strong>,  
<strong>Rqtl(1)</strong>, 
<strong>Rcross(1)</strong>, 
<strong>Qstats(1)</strong>, 
<strong>LRmapqtl(1)</strong>,
<strong>BTmapqtl(1)</strong>,
<strong>SRmapqtl(1)</strong>, 
<strong>JZmapqtl(1)</strong>, 
<strong>Eqtl(1)</strong>,
<strong>Prune(1)</strong>, 
<strong>Preplot(1)</strong>,  
<strong>MImapqtl(1)</strong>, 
<strong>MultiRegress(1)</strong>,
<strong>Examples(1)</strong>
<strong>SSupdate.pl(1)</strong>, 
<strong>Prepraw.pl(1)</strong>, 
<strong>EWThreshold.pl(1)</strong>, 
<strong>GetMaxLR.pl(1)</strong>, 
<strong>Permute.pl(1)</strong>, 
<strong>Vert.pl(1)</strong>, 
<strong>CWTupdate.pl(1)</strong>, 
<strong>Ztrim.pl(1)</strong>, 
<strong>SRcompare.pl(1)</strong>, 
<strong>Ttransform.pl(1)</strong>, 
<strong>TestExamples.pl(1)</strong>, 
<strong>Model8.pl(1)</strong>, 
<strong>Dobasics.pl(1)</strong>, 
<strong>Bootstrap.pl(1)</strong></p>
<p>
</p>
<hr />
<h1><a name="contact_info">CONTACT INFO</a></h1>
<p>In general, it is best to contact us via email (<a href="mailto:basten@statgen.ncsu.edu">basten@statgen.ncsu.edu</a>)</p>
<pre>
        Christopher J. Basten, B. S. Weir and Z.-B. Zeng
        Bioinformatics Research Center, North Carolina State University
        1523 Partners II Building/840 Main Campus Drive
        Raleigh, NC 27695-7566     USA
        Phone: (919)515-1934</pre>
<p>Please report all bugs via email to <a href="mailto:qtlcart-bug@statgen.ncsu.edu.">qtlcart-bug@statgen.ncsu.edu.</a></p>
<p>The <strong>QTL Cartographer</strong> web site ( <a href="http://statgen.ncsu.edu/qtlcart">http://statgen.ncsu.edu/qtlcart</a> ) has
links to the manual, man pages, ftp server and supplemental 
materials.</p>

</body>

</html>
